cli/main.go

```diff
@@ -0,0 +1,176 @@
+}
+
+func main() {
+	eigenpod := flag.String("eigenpod", "", "[required] The onchain address of your eigenpod contract (0x123123123123)")
+	beacon := flag.String("beacon", "", "[required] URI to a functioning beacon node RPC (https://)")
+	node := flag.String("node", "", "[required] URI to a functioning execution-layer RPC")
+	out := flag.String("output", "", "Output path for the proof. (defaults to stdout)")
+	help := flag.Bool("help", false, "Prints the help message and exits.")
+
+	flag.Parse()
+
+	if help != nil && *help {
+		flag.Usage()
+		log.Fatal("Showing help.")
+	}
+
+	if *eigenpod == "" || *beacon == "" || *node == "" {
+		flag.Usage()
+		log.Fatal("Must specify: -eigenpod, -beacon, and -node.")
+	}
+
+	ctx := context.Background()
+
+	execute(ctx, *eigenpod, *beacon, *node, out)
+}
+
+func getBeaconClient(beacon_uri string) (BeaconClient, error) {
+	beaconClient, _, err := NewBeaconClient(beacon_uri)
+	return beaconClient, err
+}
+
+func lastCheckpointedForEigenpod(eigenpodAddress string, client *ethclient.Client) uint64 {
+	eigenPod, err := onchain.NewEigenPod(common.HexToAddress(eigenpodAddress), client)
+	PanicOnError(err)
+
+	timestamp, err := eigenPod.LastCheckpointTimestamp(nil)
+	PanicOnError(err)
+	return timestamp
+}
+
+func computeSlotImmediatelyPriorToTimestamp(timestampSeconds uint64, genesis time.Time) uint64 {
+	return uint64(math.Floor(float64(timestampSeconds)-float64(genesis.Unix())) / 12)
+}
+
+// search through beacon state for validators whose withdrawal address is set to eigenpod.
+func findAllValidatorsForEigenpod(eigenpodPubKey string, beaconState *spec.VersionedBeaconState) []ValidatorWithIndex {
+	allValidators, err := beaconState.Validators()
+	PanicOnError(err)
+
+	expectedCredentials := sha256.Sum256([]byte(eigenpodPubKey))
+	expectedCredentials[0] = 0 // the first byte of withdrawal credentials is set to 0.
+
+	var outputValidators []ValidatorWithIndex = []ValidatorWithIndex{}
+	var i uint64 = 0
+	maxValidators := uint64(len(allValidators))
+	for i = 0; i < maxValidators; i++ {
+		validator := allValidators[i]
+		if validator == nil {
+			continue
+		}
+		if bytes.Equal(expectedCredentials[:], validator.WithdrawalCredentials) {
+			outputValidators = append(outputValidators, ValidatorWithIndex{
+				Validator: validator,
+				Index:     i,
+			})
+		}
+	}
+	return outputValidators
+}
+
+func getOnchainValidatorInfo(client *ethclient.Client, eigenpodAddress string, allValidators []ValidatorWithIndex) []onchain.IEigenPodValidatorInfo {
+	eigenPod, err := onchain.NewEigenPod(common.HexToAddress(eigenpodAddress), client)
+	PanicOnError(err)
+
+	var validatorInfo []onchain.IEigenPodValidatorInfo = []onchain.IEigenPodValidatorInfo{}
+
+	// TODO: batch/multicall
+	for i := 0; i < len(allValidators); i++ {
+		pubKeyHash := sha256.Sum256((allValidators[i]).Validator.PublicKey[:])
+		info, err := eigenPod.ValidatorPubkeyHashToInfo(nil, pubKeyHash)
+		PanicOnError(err)
+		validatorInfo = append(validatorInfo, info)
```

Comment:
stylistically, i prefer to do the checkpoint/withdrawal checks here and just return a list of validator indices, but not a big deal